### PR TITLE
Fix #29281 - Stop blowing in your sax if you want to sleep.

### LINKED ISSUE #29281: You can wake yourself up with the noise if you're playing an instrument when you're falling asleep.
**Describe the bug**  
You can wake yourself up with the noise if you're playing an instrument when you're falling asleep.

**To Reproduce**  
1. Get an instrument ( I specifically had this happen with a saxophone I was wearing)
2. Play the instrument
3. Without stopping, try to sleep.
4. Sometimes you'll wake up to the notice: "Something is making noise." 

**Expected behavior**  
I would expect that trying to sleep would stop you playing the instrument, as opposed to sometime around when you fall asleep.

**Versions and configuration(please complete the following information):**
 - OS: Windows 10
 - Game Version: 0.D-1102-g7c3a038
 - Graphics version: Tiles
 - Mods loaded: so many mods

### PR DESCRIPTION
**Pull request by neitsa**
#### Summary

```SUMMARY: Bugfixes "Prevent player waking up by the sound of the instrument they are playing."```

#### Purpose of change

Fixes #29281 - Prevent player waking up by the sound of the instrument they are playing.

#### Describe the solution

The problem is that the player can fall asleep while still playing an instrument and the instrument is not deactivated.

The fix deactivate the instrument in the `sleep()` method. Note that at that time in the code, the player is not sleeping (this is made later when assigning the `ACT_TRY_SLEEP` activity). Also, a message tells the user that they stopped playing the instrument.

image

#### Describe alternatives you've considered

I tried multiple things and asked myself a lot of question (like "can you really fall asleep while playing?", "at which time, if you're falling asleep, you'd stop playing?", etc.). 

I tried also removing the effects of playing an instrument (removing `playing_instrument` and `music` effects through `remove_effect()`) but that required some structural changes as the effects are located in `iuse_actor.cpp` but the sleep code is elsewhere. I wasn't also sure that removing `music` would play nicely with music players.

In the end I thought that the easiest way (and without possible side effects) was simply deactivating the instrument if the player want to sleep.

#### Additional context

Tested the fix multiple time and he worked as intended.

I'd like to thank `Julian 'Fish' Willis` who slept at total of 6 days in a basement, played the sax numerous times while being exhausted, famished and very thirsty. Your name will be remembered Julian!

### CODE REVIEW

File: src/handle_action.cpp [hunk 1]
```diff
@@ -763,6 +763,21 @@ static void sleep()
             active.push_back( mdata.name() );
         }
     }
+
+    // check for deactivating any currently played music instrument.
+    for( auto &item : u.inv_dump() ) {
+        if( item->active && item->type && !item->type->use_methods.empty() ) {
```
Comment (ifreund) [inline src/handle_action.cpp]: How about just 
```suggestion
        if( item->active && item->get_use( "musical_instrument" ) != nullptr ) {
```

**Comment by Ilysen** [comment]
The end of an era.

File: src/handle_action.cpp (hunk 1 above)
Comment (neitsa) [reply src/handle_action.cpp]: Ah thanks! didn't see this one :|

File: src/handle_action.cpp [hunk 2]
```diff
@@ -763,6 +763,21 @@ static void sleep()
             active.push_back( mdata.name() );
         }
     }
+
+    // check for deactivating any currently played music instrument.
+    for( auto &item : u.inv_dump() ) {
+        if( item->active && item->get_use( "musical_instrument" ) != nullptr ) {
+            for( const auto &method : item->type->use_methods ) {
```
Comment (ifreund) [inline src/handle_action.cpp]: Sorry, may have not been clear, you don't need this line or the next one then

**Pull request merged**